nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
-          // If argIdx being null implies the method throws an exception, then we can mark it as
-          // non-null on both non-exceptional exits from the method
-          bothUpdates.set(accessPath, Nullness.NONNULL);
+      }
+      if (!supported) {
+        // Too many arguments involved, or unsupported @Contract features. On to next clause in
+        // the
+        // contract expression
+        continue;
+      }
+      if (arg == null) {
+        // The antecedent is unconditionally true. Check for the ... -> !null case and set the
+        // return nullness accordingly
+        if (consequent.equals("!null")) {
+          return NullnessHint.FORCE_NONNULL;
         }
+        continue;
+      }
+      if (argAntecedentNullness == null) {
+        throw new IllegalStateException("argAntecedentNullness should have been set");
+      }
+      // The nullness of one argument is all that matters for the antecedent, let's negate the
+      // consequent to fix the nullness of this argument.
+      AccessPath accessPath = AccessPath.getAccessPathForNodeNoMapGet(arg, apContext);
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being false, then the return
+        // being true implies arg is not null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(
+            accessPath, Nullness.NULLABLE.greatestLowerBound(inputs.valueOfSubNode(arg)));
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being true, then the return being
+        // false implies arg is not null and we must mark it as such in the else update.
+        elseUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being true, then the return being
+        // false implies arg may be null and we must mark it as such in the else update.
+        elseUpdates.set(
+            accessPath, Nullness.NULLABLE.greatestLowerBound(inputs.valueOfSubNode(arg)));
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // Arg being non-null implies the method throws an exception, then we can mark it as
+        // null on both non-exceptional exits from the method.
+        bothUpdates.set(accessPath, Nullness.NULL);
```

Comment:
This also does not impact NullAway results, correct?

Reply:
I don't believe that this currently impacts handling (in the sections of the NullAway codebase that I'm familiar with), however it would empower other contracts if the improvement described in this comment were implemented: https://github.com/uber/NullAway/pull/669#discussion_r985768003

I'd be happy to remove this case if you like, I can't justify keeping code that I cannot confirm in a test.
I think the two best options moving forward are either:
1. Remove this update
2. Retain the update, replacing null literal checks with null flow lookups in NullEqualityVisitor, and adding a test that doesn't pass otherwise.

Reply:
I'd like to at least see a test case that would hypothetically be impacted.  Note that with `@Contract` support we are primarily (entirely?) interested in removing false positive reports, not removing false negatives.  We know NullAway is unsound in some cases and misses bugs, but I'm not too sure that working around that with `@Contract` annotations is a great idea, in terms of understandability of what the tool is doing.  I could be convinced otherwise.

Reply:
I've added a new test case which shows how this avoids false positives in unreachable code based on nullness with inline comments describing the setup: https://github.com/uber/NullAway/blob/724a5ab383d33482abe985a97f3d3807bb2cbf5c/nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java#L357-L382